### PR TITLE
Fix ParallelRunner training frequency and stale environment termination logic

### LINKED ISSUE #79: The performance decreased in parallel settings
I found that an explicitly decreased performance happened when I used the parallel setting mode in SMACv1 and SMACv2.

### PR DESCRIPTION
**Pull request by kcorder**
This PR addresses three issues in "run.py" and "parallel_runner.py" that caused significant performance degradation and potential logic errors when using parallel environments. 

### 1. Corrected training update ratio 
Previously in "run.py", the number of training updates = 1/B for each rollout phase, where B = args.batch_size_run. 
- Fixed by ensuring training frequency scales with number of episodes collected, matching behavior of standard `EpisodeRunner`. 

### 2. Stale termination mask 

In ParallelRunner, `envs_not_terminated` was computed BEFORE the env receive loop, making the select_actions() call potentially assigning values to agents incorrectly. 
- Fixed by moving mask computation after env receive. 

### 3. More efficient batches in BasicMAC 

To improve throughput, I also selected subsets of active envs for batches in `BasicMAC.forward()` (no need to update dead agents). 

**Related Issues:** 
- Possibly addresses #88 
- Fixes #79 

image

This shows that the (yellow) new parallel runs match performance of the (pink) sequential runs, 
greatly outperforming the previous (blue) parallel runs. It's faster too!

### CODE REVIEW

**Comment by LukasSchaefer** [comment]
Thanks Kevin for opening this PR, this looks great at first glance! Just wanted to give a quick sign of life that I've seen the PR. I'll have a more detailed read through the changes sometime over the coming week and will then come back to you with any follow-up questions/ concerns in case I have any!

File: src/components/episode_buffer.py [hunk 1]
```diff
@@ -86,6 +86,7 @@ def to(self, device):
 
     def update(self, data, bs=slice(None), ts=slice(None), mark_filled=True):
         slices = self._parse_slices((bs, ts))
+        slices = tuple(slices)
```
Comment (LukasSchaefer) [inline src/components/episode_buffer.py:89]: I'm not sure I follow why the type transition to `tuple` is needed here now. Could you elaborate?

File: src/components/episode_buffer.py [hunk 2]
```diff
@@ -147,6 +148,7 @@ def __getitem__(self, item):
             return ret
         else:
             item = self._parse_slices(item)
+            item = tuple(item)
```
Comment (LukasSchaefer) [inline src/components/episode_buffer.py:151]: Same here as above.

File: src/controllers/basic_controller.py [hunk 3]
```diff
@@ -19,28 +19,38 @@ def __init__(self, scheme, groups, args):
...
+        chosen_actions = self.action_selector.select_action(agent_outputs, avail_actions[bs], t_env, test_mode=test_mode)
         return chosen_actions
 
-    def forward(self, ep_batch, t, test_mode=False):
-        agent_inputs = self._build_inputs(ep_batch, t)
-        avail_actions = ep_batch["avail_actions"][:, t]
-        agent_outs, self.hidden_states = self.agent(agent_inputs, self.hidden_states)
+    def forward(self, ep_batch, t, bs=slice(None), test_mode=False):
+        # Slice the batch and hidden states to avoid running the RNNs on stale data from terminated envs
+        agent_inputs = self._build_inputs(ep_batch, t, bs=bs)
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py]: I don't think you need to modify `_build_inputs` but could just slice the batch first:
```python
sliced_ep_batch = ep_batch[bs]
agent_inputs = self._build_inputs(ep_batch, t)
avail_actions = ep_batch["avail_actions"][:, t]
```

File: src/controllers/basic_controller.py [hunk 4]
```diff
@@ -19,28 +19,38 @@ def __init__(self, scheme, groups, args):
...
+        avail_actions = ep_batch["avail_actions"][bs, t]
+
+        is_subset = not isinstance(bs, slice)
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py]: I don't follow this line.

I understand that under the default value `bs = slice(None)`, the check `isinstance(bs, slice)` is `True` and you do not effectively slice so the entire hidden state needs to be overwritten. That being said, it would be perfectly fine in that case to do
```
# bs = slice(None)
self.hidden_states[bs] = hidden_out.view_as(self.hidden_states[bs])
```
since PyTorch tensors return the entire tensor when indexed with `slice(None)`.

And when is this check `False`, what other types occur for `bs` that require other handling?

File: src/controllers/basic_controller.py [hunk 5]
```diff
@@ -19,28 +19,38 @@ def __init__(self, scheme, groups, args):
...
+        hidden_in = self.hidden_states[bs] if is_subset else self.hidden_states
+        agent_outs, hidden_out = self.agent(agent_inputs, hidden_in)
+
+        if is_subset:
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py]: This if-else case is also unclear to me but follows from the question above about `is_subset` and its use.

File: src/controllers/basic_controller.py [hunk 6]
```diff
@@ -19,28 +19,38 @@ def __init__(self, scheme, groups, args):
...
+            self.hidden_states[bs] = hidden_out.view_as(self.hidden_states[bs])
+        else:
+            self.hidden_states = hidden_out.view_as(self.hidden_states)
 
         # Softmax the agent outputs if they're policy logits
+        n_live = avail_actions.shape[0]
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py]: If you follow the suggestion to do `sliced_ep_batch = ep_batch[bs]` in the first line of this function, you don't need this variable but can just use `sliced_ep_batch.batch_size` below.

File: src/controllers/basic_controller.py [hunk 7]
```diff
@@ -60,21 +70,23 @@ def load_models(self, path):
     def _build_agents(self, input_shape):
         self.agent = agent_REGISTRY[self.args.agent](input_shape, self.args)
 
-    def _build_inputs(self, batch, t):
+    def _build_inputs(self, batch, t, bs=slice(None)):
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py]: Changes to this function should not be necessary, as per comment above.

File: src/controllers/basic_controller.py [hunk 8]
```diff
@@ -19,28 +19,38 @@ def __init__(self, scheme, groups, args):
     def select_actions(self, ep_batch, t_ep, t_env, bs=slice(None), test_mode=False):
         # Only select actions for the selected batch elements in bs
         avail_actions = ep_batch["avail_actions"][:, t_ep]
-        agent_outputs = self.forward(ep_batch, t_ep, test_mode=test_mode)
-        chosen_actions = self.action_selector.select_action(agent_outputs[bs], avail_actions[bs], t_env, test_mode=test_mode)
+        agent_outputs = self.forward(ep_batch, t_ep, bs=bs, test_mode=test_mode)
```
Comment (LukasSchaefer) [inline src/controllers/basic_controller.py:22]: I generally like this change to avoid wasteful compute. That being said, I would like to be explicit about any assumptions being made here.

As I understand, for this change you assume hat within each episode, any indexing that filters out particular agents/ environments through `bs` will continue to filter out that agent/ environment for the rest of the episode (e.g. agents "dying" in SMAC or parallel environments being terminated already). If this assumption is ever violated, this change causes issues since only stepping forward for particular agents updates only their hidden state, i.e. agents that have not been "stepped" will have wrong/ outdated hidden states for the remainder of the episode.

Is my understanding about assumptions you are making correct?

File: src/run.py [hunk 9]
```diff
@@ -201,16 +201,19 @@ def run_sequential(args, logger):
         buffer.insert_episode_batch(episode_batch)
 
         if buffer.can_sample(args.batch_size):
-            episode_sample = buffer.sample(args.batch_size)
+            n_updates = args.batch_size_run
```
Comment (LukasSchaefer) [inline src/run.py]: I don't agree with this "fix". It's not a bug but intentional design for many online algorithms like COMA or MAPPO to gather data by interacting with multiple parallel environments and then updating on the batch of all these parallel environments. This change would mean that MAPPO e.g. would collect episodes from 10 parallel environments to then do 10 updates in a row using the same batch from these 10 parallel environments. While it can be sensible to do multiple updates, I would consider this decision non-standard and would not want to incorporate it into EPyMARL by default.

Please let me know if you disagree with my assessment, happy to discuss if this could e.g. be a sensible option (perhaps introducing a new config parameter like `batch_num_updates` or something like that).

File: src/runners/parallel_runner.py [hunk 10]
```diff
@@ -154,23 +154,19 @@ def run(self, test_mode=False):
...
-            if all_terminated:
-                break
-
             # Post step data we will insert for the current timestep
             post_transition_data = {"reward": [], "terminated": []}
             # Data for the next step we will insert in order to select an action
             pre_transition_data = {"state": [], "avail_actions": [], "obs": []}
 
+            # Track which envs contributed data this step (alive before receive)
+            envs_that_stepped = []
```
Comment (LukasSchaefer) [inline src/runners/parallel_runner.py:163]: I believe that this will be identical to the `envs_not_terminated` defined before in the original code since both accumulate the indices of non-terminated environments as per the `terminated` variable.

File: src/runners/parallel_runner.py [hunk 11]
```diff
@@ -194,22 +190,32 @@ def run(self, test_mode=False):
...
+            envs_not_terminated = [
+                b_idx for b_idx, termed in enumerate(terminated) if not termed
+            ]
+            all_terminated = all(terminated)
+
+            # Post-transition data goes to ALL envs that stepped (includes just-terminated)
             self.batch.update(
                 post_transition_data,
-                bs=envs_not_terminated,
+                bs=envs_that_stepped,
```
Comment (LukasSchaefer) [inline src/runners/parallel_runner.py:202]: As per my comment above, I believe this does not change the code semantics.

File: src/runners/parallel_runner.py [hunk 12]
```diff
@@ -194,22 +190,32 @@ def run(self, test_mode=False):
...
 
             # Move onto the next timestep
             self.t += 1
 
-            # Add the pre-transition data
+            # Pre-transition data also goes to all envs that stepped, so their
+            # terminal observations are stored (needed by the learner for target Q)
             self.batch.update(
-                pre_transition_data, bs=envs_not_terminated, ts=self.t, mark_filled=True
+                pre_transition_data, bs=envs_that_stepped, ts=self.t, mark_filled=True
```
Comment (LukasSchaefer) [inline src/runners/parallel_runner.py:213]: Same as above, I don't think this actually changes anything.

File: src/runners/parallel_runner.py [hunk 13]
```diff
@@ -154,23 +154,19 @@ def run(self, test_mode=False):
                     if idx == 0 and test_mode and self.args.render:
                         parent_conn.send(("render", None))
 
-            # Update envs_not_terminated
```
Comment (LukasSchaefer) [inline src/runners/parallel_runner.py:157]: I think this timing is actually sensible and correct. For the `bs` indexing, as per my comments below, I don't think that the new variable `envs_that_stepped` represents any functional change. Additionally, the break is placed here since environments that terminate still return their next observation, available actions, and state to which the code above is storing the actions. This can be used in updates for the last final step and should be stored.

**Review by LukasSchaefer — COMMENTED** [review]
Hi @kcorder, thanks a lot for creating the PR with your suggestions, we greatly appreciate any contributions!

I had a look and am not sure that I agree with all the suggested changes. Please see my more detailed in-line comments below. I'm happy to iterate/ discuss if you disagree with my assessment.

File: src/components/episode_buffer.py (hunk 1 above)
Comment (kcorder) [reply src/components/episode_buffer.py:89]: This is to suppress PyTorch warnings in newer versions (the versions are not pinned in "requirements.txt"). I have version 2.8.0.

File: src/components/episode_buffer.py (hunk 2 above)
Comment (kcorder) [reply src/components/episode_buffer.py:151]: Same warning suppress for future deprecation

File: src/run.py (hunk 9 above)
Comment (kcorder) [reply src/run.py]: That makes sense to add a separate config for parallel runners - it is technically changing the algorithm a bit, and good catch! That will affect the PG algorithms more directly because they only have the rollout buffer, not a large experience buffer to sample batches from. Maybe `num_updates_per_rollout` would be a good abstraction to enable parallel for the value-based algorithms? 

I'll add MAPPO as a test case along with QMIX for whatever is changed.

File: src/runners/parallel_runner.py (hunk 13 above)
Comment (kcorder) [reply src/runners/parallel_runner.py:157]: This reply applies to the other `envs_that_stepped` and `envs_not_terminated` comments too: 
I traced through this for a while and I agree with you now - they are functionally the same wrt training. 

Although it does have a performance benefit (and slightly cleaner logic in my opinion) by separating them like this. Even though it doesn't affect policies, the terminated envs are still currently passed to the next iteration into `select_actions()` unnecessarily. Unless there's another algorithm edge case I'm not thinking of here.

File: src/controllers/basic_controller.py (hunk 3 above)
Comment (kcorder) [reply src/controllers/basic_controller.py]: Good call this is cleaner

File: src/controllers/basic_controller.py (hunk 4 above)
Comment (kcorder) [reply src/controllers/basic_controller.py]: That is also true and I didn't catch it. The one difference is that indexing returns a view, but shouldn't be an issue in torch RNNs that don't modify state in-place. In my new commit I reverted this `is_subset` stuff.

**Comment by kcorder** [comment]
Hi, I have reverted most of these changes per your comments. I added a `num_updates_per_rollout` arg for the update ratio, with user warnings if they're using non-recommended settings. 

Here are quick benchmark results including MAPPO (green) post-changes: 
image

File: README.md [hunk 14]
```diff
@@ -232,6 +232,16 @@ EPyMARL defines yaml configuration files for algorithms and environments under `
...
 
+### `num_updates_per_rollout`
+
+The `num_updates_per_rollout` option (default: `1`) controls how many gradient updates are performed after each environment collection phase. For **off-policy, value-based algorithms** (e.g. QMIX, VDN, IQL) that use a large replay buffer, using parallel rollouts with batched updates can greatly improve training speed at the cost of some sample efficiency. 
+
+```sh
+python src/main.py --config=qmix --env-config=smaclite with env_args.time_limit=150 env_args.map_name="MMM" batch_size_run=30 num_updates_per_rollout=30 runner="parallel" 
+```
+
+**Do not use this with policy-gradient algorithms** (MAPPO, IPPO, IA2C, MAA2C, COMA, PAC). Their buffers hold exactly one rollout (`buffer_size == batch_size`), so repeated `sample()` calls return identical data.
```
Comment (LukasSchaefer) [inline README.md]: "Their buffers hold exactly one rollout" doesn't sound quite accurate. I'd suggest: "Their buffers contain only the most recently collected on-policy episodes".

**Comment by LukasSchaefer** [comment]
> Hi, I have reverted most of these changes per your comments. I added a `num_updates_per_rollout` arg for the update ratio, with user warnings if they're using non-recommended settings.
> 
> Here are quick benchmark results including MAPPO (green) post-changes: image

Just to understand these shared results, which run corresponds to which setting? Are they using the multiple updates per episode variable? Where are the speedups mainly coming from?

**Comment by kcorder** [comment]
This is a mix - the pink sequential runs and (old) blue QMIX parallel runs are the same as in the previous screenshot. This figure shows MAPPO (green) and QMIX (yellow) using the newest code. 
MAPPO needs no config changes from default (already used parallel runs), and QMIX was run with 
`batch_size_run=30 num_updates_per_rollout=30 runner="parallel"`. 
It throws a warning when using a value-based algorithm (replay, not rollout buffer) with parallel runner and not setting `batch_size_run == num_updates_per_rollout`. 

The speedup is pretty minor, but is more significant with more parallel envs (and bigger rollout length variance in batches). It comes from the `envs_not_terminated` re-compute after the env receive here: https://github.com/kcorder/epymarl-parallel-rnn-fix/blob/6e6c6bc19c1a7f2cd85444b93985c1ebaa9d260e/src/runners/parallel_runner.py#L193-L197  
Then at the next loop iteration, the MAC can correctly mask out the newly terminated envs from the batch instead of forward() passing and throwing away the result.

**Comment by kcorder** [comment]
Hi @LukasSchaefer, how should this proceed?  I sorta forgot about it, but my teammates have been using this PR.
